### DISCUSSION TITLE
Self Hosting 3rd party registries & adding/removing 3rd party registries into composer

### DISCUSSION THREAD

**Discussion by trymeouteh** [composer/composer#11227] [General]
- The ability to self host your own composer repo which will require releasing composer to be fully open source and instructions on how to deploy your own composer registry to self host your own package and optionally allow others to host their packages on the 3rd party registry.
- The ability to add/remove registries in composer. This will require new commands such ad _addrepo_, _removerepo_ and _listrepos_.
- The ability to download a package from a 3rd party registry. the require command must allow domain names such as _composer require example.com:moment/moment_.
- The ability to have a default registry set and the ability to change the default registry. By default the default registry will be packagist.org, the official registry but to have the ability to change this with a command such as _changedefaultrepo_. The user is able to download packages from the default registry without having to put the domain name in front of it, for example if the default registry is packagist.org, the user can enter _composer require moment/moment_ and not need to enter _composer require example.com/moment/moment_.
- The ability to have packages require dependencies from packages on different registries.

Such a feature will allow for decentralisation, private registries and the ability to truly self host your own packages.

**Comment by stof** [comment]
Composer already supports using several repositories for metadata. However, package names are not considered as being tied to a particular repository (unlike your `composer require example.com:moment/moment` suggestion). But this means that cross-repository dependencies already work.
Btw, this support for extra repositories without a separate package namespacing is precisely what allows to use a fork of a package or to rely on Private Packagist as a proxy of packagist.org for instance.

In Composer 2.x, registered repositories have a priority, which still gives you control about which repository can be the source for a given package name: https://getcomposer.org/doc/articles/repository-priorities.md

> * The ability to add/remove registries in composer. This will require new commands such ad _addrepo_, _removerepo_ and _listrepos_.

This is covered already by the `composer config` command.

**Reply by trymeouteh** [reply]
Do you know of any 3rd party registries that I can add into composer and try installing packages from?
